README.md
# express-mw-lang

[![Build Status](https://travis-ci.org/volebo/express-mw-lang.svg?branch=master)](https://travis-ci.org/volebo/express-mw-lang)
[![codecov](https://codecov.io/gh/volebo/express-mw-lang/branch/master/graph/badge.svg)](https://codecov.io/gh/volebo/express-mw-lang)
[![bitHound Overall Score](https://www.bithound.io/github/volebo/express-mw-lang/badges/score.svg)](https://www.bithound.io/github/volebo/express-mw-lang)
[![bitHound Dependencies](https://www.bithound.io/github/volebo/express-mw-lang/badges/dependencies.svg)](https://www.bithound.io/github/volebo/express-mw-lang/master/dependencies/npm)
[![npm version](https://img.shields.io/npm/v/express-mw-lang.svg)](https://www.npmjs.com/package/express-mw-lang)
[![npm downloads](https://img.shields.io/npm/dm/express-mw-lang.svg)](https://www.npmjs.com/package/express-mw-lang)

Localization helper for Express web server.

Cuts off the **language** part from the path, and allow to use simple routes!

This middleware helps to determine the language, and handles urls of incoming requests for next middlewares. `express-mw-lang` does **not** use `Accept-Language` (but it is a good idea, to use the value of this HTTP-Header on first contact with the user, to predict his language).

To determine the language `express-mw-lang` uses only **URL** of the request. Expected format of the URL:

```
schema://host/< LANG >/path?params
```

Sample URL, handled in French:

```
https://example.net/fr/article/1
```

As you could see - `lang` is a first part of the path, and could be ommited, in this case the **default** language will be used.

Actually, `express-mw-lang` **acts and is** an [express-router](http://expressjs.com/en/4x/api.html#router). The best practice is appending **your** router/routes (`use`/`route`/`get`/`post`/...) to the `express-mw-lang`, instead of habitual way with the Express-application itself.

See the [Example section](#examples) for more details.

[Contributions are welcome][contributing]!

## Usage

Install:

```bash
$ npm install express-mw-lang
```

Configure and use in the `app.js`:

```javascript
// ...
var langGen = require('express-mw-lang');
var app = express();

// ...

let langmw = langGen();
langmw.esu(app);

```

## Example

```javascript
"use strict";

var express         = require('express');
var logger          = require('morgan');          // just for example, not required
var bodyParser      = require('body-parser');     // just for example, not required
var langGen         = require('express-mw-lang');

var app = express();

app.use(logger());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

// STEP 1: CREATING the middleware instance with options:
var langmw = langGen({
	defaultLanguage: 'en',
	availableLanguages: ['en', 'ru'],
});
// STEP 2: append middleware to application
// ( `esu` - is a reverse of `use`, because app and mw are swapped):
langmw.esu(app);


// appending main routes to the app (through lang-mw):
var router = express.Router();
router
	.route('/home')
	.get(function(req, res, next){
		res.status(200).send('Hello [' + req.lang.code + ']');
	});
// STEP 3: append your routes to lang-mw
langmw.use(router);

app.listen(3000, function() {
	console.log('Example app listening on port 3000!');
});

// http://192.168.1.68:3000/en/home
// -> Hello [en]
// http://192.168.1.68:3000/ru/home
// -> Hello [ru]
// http://192.168.1.68:3000/home
// -> Hello [en]
```

## Options

```javascript
const options = {
	defaultLanguage: 'en',
	availableLanguages: [],
	onLangDetected: function(langCode, req, res) {}
};
```

#### `defaultLanguage`

Default: **`'en'`**, _string ([ISO 639](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes))_.

Default language

#### `availableLanguages`

Default: **`[]`**, _array of strings (each string - ISO 639 code)_.

Languages, which will be recognized by middleware.

1. Default language is always available
2. For requests to unavailable language the response will contain redirect to the `/` of the site.

#### `onLangDetected`

Defalt: **`null`**, _function(langCode, req, res)_

Callback, called when the lang is determined. Convenient place to setup the locale for `momentjs` or choosen i18n library.

##### Example

```javascript

var i18n            = require("i18n")
var moment          = require('moment')

// ...

let options = {
	defaultLanguage: 'en',
	availableLanguages: ['en', 'ru'],
	onLangDetected: function(lang_code, req, res) {

		i18n.setLocale(lang_code)
		i18n.setLocale(req, lang_code)
		i18n.setLocale(res, lang_code)

		moment.locale(lang_code)
	}
}
```

## `req` and `res` extensions

This middleware extends the `req` and `res` objects with a `lang` property, so extensions are:

1. `req.lang`
2. `res.locals.lang`

Further - description of the `lang` object.

#### `defaultLanguage`

_string_, default language, taken from options.

#### `available`

_array_ of available options. Each item is an object with `code`-property (string, ISO 639)

#### `usingDefault`

_bool_ indicates, that request is handled using default language (probably, there was an error, when the MW tried to determine the required language of the request).

#### `routeTo`

_function(local_path, lang_code)_ , this function could help you to build paths to other pages of the site (taking the current language into account).

Notes about argument `lang_code`:

1. it is optional
2. UNKNOWN/UNAVAILABLE lang_code - cause to use the default lang
3. NULL lang_code - cause to use CURRENT lang (determined by MW from URL)
4. EMPTY ('') lang_code - cause to use default lang!!

## Contributing

You could take part in the development process, just follow this [guideline][contributing].

## License

Please, read the [`LICENSE`](LICENSE) file in the root of the repository (or downloaded package).

[contributing]: CONTRIBUTING.md


CHANGELOG.md
# CHANGELOG

```yaml
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################
```

# CHANGELOG

This is the history of changes of the `volebo/express-mw-lang` project.

This file **MUST** be filled **ONLY** by maintainers, using messages from pull
requests.

Please, follow this [guide](http://keepachangelog.com/en/0.3.0/).

## unpublished // ??

## 2.0.0 // 2017-10-28

* workaround (or hack) for problem with `/me` URLs (previoulsy they lead to 404)
* throw Errors on invalid options (like a wrong `defaultLanguage`)

* rename `onLangCodeReady` => `onLangDetected`
* expose `available` and `defaultLanguage` properties from the created Router object

* update all deps
* add changelog

## 1.3.2 // ???

Evolution without changelogs...


test/.eslintrc.js
exports = module.exports = {
	"env": {
		"mocha": true,
	},
	"globals": {
		"packageRoot": true,
		"expect": true
	}
}


test/bootstrap.js
'use strict'

const chai            = require('chai')
const mh              = require('@volebo/mocha-helpers')

global.expect = chai.expect
global.filename2suitename = mh.filename2suitename

/* ROOT of the package */
global.packageRoot = process.cwd()


test/detect-lang.js
/*
Language-helper middleware for Express web server.

Copyright (C) 2016  Volebo.Net <volebo.net@gmail.com>
Copyright (C) 2016  Koryukov Maksim <maxkoryukov@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the MIT License, attached to this software package.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You should have received a copy of the MIT License along with this
program. If not, see <https://opensource.org/licenses/MIT>.
*/

"use strict";

const request  = require('supertest');
//const assert   = require('chai').assert;
//const path     = require('path');
const express  = require('express');
//const debug    = require('debug')('volebonet:express:mw:lang:test');

/* ROOT of the package */
var rt = process.cwd();

describe('Language detection', function(){

	let app = express();
	let langmw = {};
	const deflang = 'en';

	before(function() {
		langmw = require(rt)({
			defaultLanguage: deflang,
			availableLanguages: ['en', 'ru', 'zh-CHS'],
		});
		langmw.esu(app);
	});

	describe('should return explicit lang code (from available list)', function () {
		// set up handler
		before(function() {
			langmw.get('/code', (req, res, next) => {
				res.status(200).send(res.locals.lang.code);
				next();
			});
		});

		[
			// to check default twice
			{ path: '/code', resp: 'en' },

			{ path: '/ru/code', resp: 'ru' },
			{ path: '/en/code', resp: 'en' },
			{ path: '/zh-chs/code', resp: 'zh-CHS' },
		].forEach( data => {

			it(data.path, function(done) {
				request(app)
					.get(data.path)
					.expect(200, data.resp, done);
			});
		});
	});

	describe('should return default lang code', function() {
		// set up handler
		before(function() {
			langmw.get('/default', (req, res, next) => {
				res.status(200).send(res.locals.lang.defaultLanguage);
				next();
			});
		});

		[
			{ path: '/default' },
			{ path: '/ru/default' },
			{ path: '/en/default' },
			{ path: '/zh-chs/default' },
		].forEach( data => {

			it(data.path, function(done) {
				request(app)
					.get(data.path)
					.expect(200, deflang, done);
			});
		});
	});

	describe('should RUN NEXT MW on unknown language/culture', function() {

		const RESP_NO_LANG = 'looks like a lang';
		const RESP_VK = 'vk response';

		const PATH_NO_LANG = '/xx-YY/looks-like-lang-but-no';
		const PATH_VK = '/vk/auth';
		const PATH_VK_NO = '/vk/no-such-page';

		before(function() {
			// appending next middlewares
			langmw.get(PATH_NO_LANG, (req, res, next) => {
				res.status(200).send(RESP_NO_LANG);
				next();
			});

			langmw.post(PATH_VK, (req, res, next) => {
				res.status(200).send(RESP_VK);
				next();
			});

		});

		it(PATH_NO_LANG + ' and should not eat the prefix', function(done) {
			request(app)
				.get(PATH_NO_LANG)
				.expect(200, RESP_NO_LANG, done);
		});

		[
			PATH_VK,
			'/ru' + PATH_VK,
			'/en' + PATH_VK
		].forEach(function(urlpath){
			it(urlpath + ' and should not eat the prefix', function(done) {
				request(app)
					.post(urlpath)
					.expect(200, RESP_VK, done);
			});
		});

		[
			PATH_VK_NO,
			'/ru' + PATH_VK_NO,
			'/en' + PATH_VK_NO
		].forEach(function(urlpath){
			it(urlpath, function(done) {
				request(app)
					.get(urlpath)
					.expect(404, done);
			});
		});

	});

	// behaviour changed, see https://github.com/VoleboNet/express-mw-lang/issues/8
	describe('should RUN NEXT MW for unavailable lang NOT TRIMMING the url', function () {
		// set up handler
		before(function() {

			// the MW will not eat the LANG PREFIX (because it is not
			// available), so we should add handler for the full path:
			langmw.get('/ru-ru/code-def-unavail-lang', (req, res, next) => {
				res.status(200).send(res.locals.lang.code);
				next();
			});

			langmw.get('/zh/code-def-unavail-lang', (req, res, next) => {
				res.status(200).send(res.locals.lang.code);
				next();
			});

		});

		[
			{ path: '/ru-ru/code-def-unavail-lang' },
			{ path: '/ru-RU/code-def-unavail-lang' },
			{ path: '/zh/code-def-unavail-lang' },
		].forEach( data => {

			it(data.path, function(done) {
				request(app)
					.get(data.path)
					.expect(200, deflang, done);
			});
		});


		[
			{ path: '/ru-ru/code-unavail-lang-404' },
			{ path: '/ru-RU/code-unavail-lang-404' },
			{ path: '/zh/code-unavail-lang-404' },
		].forEach( data => {

			it(data.path, function(done) {
				request(app)
					.get(data.path)
					.expect(404, done);
			});
		});
	});

});


.eslintrc.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################

Language-helper middleware for Express web server.

Copyright (C) 2016-2017 Volebo <dev@volebo.net>
Copyright (C) 2016-2017 Maksim Koryukov <maxkoryukov@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the MIT License, attached to this software package.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You should have received a copy of the MIT License along with this
program. If not, see <https://opensource.org/licenses/MIT>.
*/

const path = require('path')

exports = module.exports = {
	'extends': [
		// 'volebo',
		path.join(__dirname, 'node_modules', 'eslint-config-volebo', 'index.js')
	],

	'overrides': [
		{
			'files': [
				'**/*.test.js',
				'**/*.spec.js'
			],

			'globals': {
				'filename2suitename': true,
			}
		}
	],
}


src/known-langs.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################

Language-helper middleware for Express web server.

Copyright (C) 2016-2017 Volebo <dev@volebo.net>
Copyright (C) 2016-2017 Maksim Koryukov <maxkoryukov@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the MIT License, attached to this software package.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You should have received a copy of the MIT License along with this
program. If not, see <https://opensource.org/licenses/MIT>.
*/

'use strict'

const _ = require('lodash')

// TODO : load from ICU
const _langs = [
	{
		// NOTE: do not remove the EN locale, it is library default!
		// https://en.wikipedia.org/wiki/IETF_language_tag
		code: 'en',
		name: {
			short: 'en',
			full: 'English',
			native: {
				short: 'en',
				full: 'English'
			}
		},
	},

	{
		code: 'en-GB',
		name: {
			short: 'en (GB)',
			full: 'English (GB)',
			native: {
				short: 'en (GB)',
				full: 'English (GB)'
			}
		},
	},

	{
		code: 'en-US',
		name: {
			short: 'en (US)',
			full: 'English (US)',
			native: {
				short: 'en (US)',
				full: 'English (US)'
			}
		},
	},

	{
		code: 'ru',
		name: {
			short: 'ru',
			full: 'Russian',
			native: {
				short: 'рус',
				full: 'Русский'
			}
		},
	},

	{
		code: 'ru-RU',
		name: {
			short: 'ru (RU)',
			full: 'Russian (Russia)',
			native: {
				short: 'рус (Рос)',
				full: 'Русский (Россия)'
			}
		},
	},

	{
		code: 'zh',
		name: {
			short: 'zh',
			full: 'Chinese',
			native: {
				short: '吉恩斯',
				full: '吉恩斯'
			}
		},
	},
	{
		code: 'zh-CHS',
		name: {
			short: 'zh-CHS',
			full: 'Chinese (Simplified)',
			native: {
				short: '那么-ES',
				full: '吉恩斯（简体）'
			}
		},
	},
]

const _dic = _.keyBy(_langs, v => v.code.toLowerCase())

module.exports = _dic


src/index.js
/*
################################################################################
#                                                                              #
# db    db  .8888.  dP     888888b 8888ba   .8888.     d8b   db 888888b d8888P #
# 88    88 d8'  `8b 88     88      88  `8b d8'  `8b    88V8  88 88        88   #
# Y8    8P 88    88 88    a88aaa   88aa8P' 88    88    88 V8 88 88aaa     88   #
# `8b  d8' 88    88 88     88      88  `8b 88    88    88  V888 88        88   #
#  `8bd8'  Y8.  .8P 88     88      88  .88 Y8.  .8P dP 88   V88 88        88   #
#    YP     `888P'  88888P 888888P 888888'  `888P'  88 VP    8P 888888P   dP   #
#                                                                              #
################################################################################

Language-helper middleware for Express web server.

Copyright (C) 2016-2017 Volebo <dev@volebo.net>
Copyright (C) 2016-2017 Maksim Koryukov <maxkoryukov@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the MIT License, attached to this software package.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

You should have received a copy of the MIT License along with this
program. If not, see <https://opensource.org/licenses/MIT>.
*/

'use strict'

const debug           = require('debug')('volebo:express:mw:lang')
const _               = require('lodash')
const express         = require('express')
const deepFreeze      = require('deep-freeze')

const urlBuilder      = require('url').parse

const knownLangs = require('./known-langs')
// dict with langs structures,
//
//	 keys   : ISO 639-x codes (lowercase)
//	 values : looks like this:
//
//	 {
//	 	code: 'en',
//	 	name: {
//	 		short: 'en',
//	 		full: 'English',
//	 		native: {
//	 			short: 'en',
//	 			full: 'English'
//	 		}
//	 	}
//	 }


const LangMw = function(options) {

	const _opt = options || {}

	let defLangCode = _.toString(_opt.defaultLanguage) || 'en'
	const defLangData = knownLangs[_.toLower(defLangCode)]

	if (!defLangData) {
		throw new Error(`Unknown defaultLanguage: [${defLangCode}]`)
	}

	defLangCode = defLangData.code

	const available_lang_codes = _opt.availableLanguages || []
	available_lang_codes.push(defLangCode)

	const _available = _(available_lang_codes)
		.uniq()
		.map(lcode => knownLangs[_.toLower(lcode)])
		.filter()
		.value()
	const availableFrozen = deepFreeze(_available)
	const _availableDict = _.keyBy(availableFrozen, v => _.toLower(v.code))

	const _onLangDetected = _opt.onLangDetected || _.noop

	if (!_.isFunction(_onLangDetected)) {
		throw new TypeError('options.onLangDetected MUST be a function')
	}

	// -------------------------------------------------------------------------
	// OPTIONS PARSED
	// -------------------------------------------------------------------------
	// debug('defaultLanguage', defLangCode)
	// debug('availableFrozen', availableFrozen)

	const mw_lang_pre_handler = function mw_lang_pre_handler(req, res, next) {
		let lang_code = _.get(req.params, 'lang')
		let is_defLangCode

		if (_.isNil(lang_code)) {
			is_defLangCode = true
			lang_code = defLangCode
		} else {
			is_defLangCode = false
			const cult_code = _.get(req.params, 'cult', '')
			lang_code = lang_code + cult_code
		}

		let lc = _availableDict[_.toLower(lang_code)]

		if (!lc){
			// TODO: is this solution slow?

			const _urlPathname = urlBuilder(req.originalUrl).pathname

			// if the `originalUrl` ends with a slash - everything will work
			// just fine.
			// but if the `originalUrl` is something like [/me] or [/vk-ME]
			// than express-url-trimmer will do bad work for us.
			// To avoid this it is enough to add an additional slash at the
			// begging of req.url
			//
			// HACK: but this solution could be considered as a hack!
			//
			const _prependSlash = '/' !== _urlPathname[_urlPathname.length-1]

			const _newUrl = _prependSlash ? '/' + req.originalUrl : req.originalUrl
			debug('use default culture and restore URL to original value [%s]', _newUrl)
			req.url = _newUrl

			lc = defLangData
			is_defLangCode = true
		}

		const localeinfo = {}

		localeinfo.defaultLanguage = defLangCode
		localeinfo.available = availableFrozen
		localeinfo.usingDefault = is_defLangCode

		localeinfo.routeTo = function _routeTo(local_path, explicit_lang_code) {
			if (_.isNil(explicit_lang_code)) {
				explicit_lang_code = localeinfo.code
			}

			// TODO: it could be precalculated
			// validate explicit_lang_code:
			const lang = _availableDict[_.toLower(explicit_lang_code)]
			const effective_lang_code = lang ? lang.code : defLangCode

			let p = local_path
			if (effective_lang_code !== defLangCode) {
				p = `/${effective_lang_code}${local_path}`
			}

			return p
		}

		localeinfo.setLocale = function _setLocale(lc) {

			const new_lc = _availableDict[_.toLower(lc)]

			if (!new_lc) {
				throw new Error(`Locale not found: ${lc}`)
			}

			_.assign(this, new_lc)

			_onLangDetected(new_lc.code, req, res)
		}

		// Set up req and res:
		res.locals.lang = localeinfo
		req.lang = localeinfo

		localeinfo.setLocale(lc.code)

		return next()
	}

	const _router = new express.Router()
	_router.available = availableFrozen
	_router.defaultLanguage = defLangCode

	_router.esu = function(app) {
		//
		app.use('/:lang(\\w{2})?:cult([-_]\\w{2,3})?', mw_lang_pre_handler, _router)
	}

	return _router
}

exports = module.exports = LangMw
